x509util/crlcheck/crlcheck.go

```diff
@@ -19,67 +19,203 @@ package main
-			crls = append(crls, certList)
-			if *verbose {
-				fmt.Print(x509util.CRLToString(certList))
-			}
+			glog.Infof("treating cert [%d] with subject %q as potential issuer", i, x509util.NameToString(issuer.Subject))
+			caCerts = append(caCerts, issuer)
 		}
-		for _, caCert := range caCerts {
-			for _, certList := range crls {
-				if err := caCert.CheckCertificateListSignature(certList); err != nil {
-					fmt.Fprintf(os.Stderr, "%s: Verification error: %v\n", filename, err)
-					errcount++
+	}
+	return processCert(dataList[0], caCerts)
+}
+
+func processCert(data []byte, caCerts []*x509.Certificate) error {
+	client := &http.Client{}
+
+	cert, err := x509.ParseCertificate(data)
+	if err != nil {
+		return fmt.Errorf("certificate parse error: %v", err)
+	}
+	issuer, err := getIssuer(cert, client)
+	if err != nil {
+		glog.Warningf("failed to retrieve issuer for cert: %v", err)
+	}
+	if issuer != nil {
+		glog.Infof("using issuer %q", x509util.NameToString(issuer.Subject))
+		caCerts = append(caCerts, issuer)
+	}
+	expired := false
+	if time.Now().After(cert.NotAfter) {
+		glog.Errorf("certificate is expired (since %v)", cert.NotAfter)
+		expired = true
+	}
+	for _, crldp := range cert.CRLDistributionPoints {
+		glog.Infof("retrieving CRL from %q", crldp)
+		rsp, err := client.Get(crldp)
+		if err != nil || rsp.StatusCode != http.StatusOK {
+			return fmt.Errorf("failed to get CRL from %q: %v", crldp, err)
+		}
+		body, err := ioutil.ReadAll(rsp.Body)
+		if err != nil {
+			return fmt.Errorf("failed to read CRL from %q: %v", crldp, err)
+		}
+		rsp.Body.Close()
+		certList, err := processCRL(body, caCerts)
+		if err != nil {
+			return err
+		}
+		if expired {
+			continue
+		}
+		// Check the CRL for the presence of the original cert.
+		for _, rev := range certList.TBSCertList.RevokedCertificates {
+			if rev.SerialNumber.Cmp(cert.SerialNumber) == 0 {
+				glog.Errorf("%s: certificate with serial number %v revoked at %v", crldp, cert.SerialNumber, rev.RevocationTime)
+				if rev.RevocationReason != x509.Unspecified {
+					glog.Errorf("  revocation reason: %s\v", x509util.RevocationReasonToString(rev.RevocationReason))
 				}
+				break
 			}
 		}
 	}
-	if errcount > 0 {
-		os.Exit(1)
+
+	return nil
+}
+
+func getIssuer(cert *x509.Certificate, client *http.Client) (*x509.Certificate, error) {
+	if len(cert.IssuingCertificateURL) == 0 {
+		return nil, nil
+	}
+	issuerURL := cert.IssuingCertificateURL[0]
+	glog.Infof("retrieving issuer from %q", issuerURL)
+	rsp, err := client.Get(issuerURL)
+	if err != nil || rsp.StatusCode != http.StatusOK {
+		return nil, fmt.Errorf("failed to get issuer from %q: %v", issuerURL, err)
+	}
+	body, err := ioutil.ReadAll(rsp.Body)
+	if err != nil {
+		return nil, fmt.Errorf("failed to read issuer from %q: %v", issuerURL, err)
+	}
+	rsp.Body.Close()
+	issuers, err := x509.ParseCertificates(body)
+	if err != nil {
+		return nil, fmt.Errorf("failed to parse --issuer cert: %v", err)
 	}
+	return issuers[0], nil
```

Comment:
worth caching these?

Reply:
That feels like premature optimization (until we're in a situation where this tool is being applied to large sets of certs from the same issuer).

Reply:
fair enough